### DISCUSSION TITLE
Any tips for setting up vscode for native module development?

### DISCUSSION THREAD

**Discussion by lerouxb** [micropython/micropython#13596] [General]
Starting with a new project that's just the basic factorial natmod example from here https://docs.micropython.org/en/latest/develop/natmod.html

I have that open in a workspace in vscode. I'm getting a lot of `expression must have struct or union type but it has type "const int"` errors. I'm wondering if anyone's done the work of configuring vscode to work with this?

<img width="948" alt="Screenshot 2024-02-04 at 16 03 39" src="https://github.com/micropython/micropython/assets/69737/b6c10e85-10b0-44eb-b923-ce3beaf8803d">

I found the vscode page on the wiki https://github.com/micropython/micropython/wiki/VS-Code but it doesn't seem to deal with setting things up for c.

I'd love some pointers if possible.

**Comment by jimmo** [comment]
@lerouxb The problem here is that whatever is powering the vscode popup doesn't understand the magic of where `mp_fun_table` comes from, because it doesn't know which preprocessor flags need to be set. (in this case `MICROPY_ENABLE_DYNRUNTIME`)

See the end of `py/nativeglue.h` -- it sees `extern const int mp_fun_table;` but it should see `extern const mp_fun_table_t mp_fun_table;`.

I'm not sure the best way to fix this... are you able to configure vscode to know about the following flags that `py/dynruntime.mk` sets:

```
CFLAGS += -DNO_QSTR
CFLAGS += -DMICROPY_ENABLE_DYNRUNTIME
CFLAGS += -DMP_CONFIGFILE='<$(CONFIG_H)>'
```
